### ISSUE TITLE
Synchronization between components

### ISSUE THREAD

**Issue by nanmuxx** [simbricks/simbricks#48]
After reading the SimBricks paper, I still don't understand how synchronization between components works. Can you give me some more details and what part of the source code this corresponds to? There seems to be no specific introduction to the adapter. Could you give me some details as well? Looking forward to your reply.

**Comment by hajeongee** [comment]
Hi @nanmuxx 
The essential of synchronization is,
a simulator only proceeds to time T only when it receives messages with arrival time T or larger from all the channels.
Or in other words, the simulator checks the received message timestamp for all the channels it has, and proceeds to the smallest timestamp of them. If it already has proceeded to the timestamp, it will busy wait at the queue until there is a new message arrives and updates the smallest timestamp.

NIC simulator as an example, it has two channels: Ethernet and PCIe.
Suppose NIC received a message from Ethernet channel with timestamp T, and a message from PCIe with timestamp T + 5.
Then NIC simulator is allowed to advance to time T.
Now that NIC's local clock is T, and it received a new message from Ethernet channel with timestamp T + 3.
Then it will choose the smallest timestamp from the two channels, which is T + 3 to advance. 

You can refer to [this](https://github.com/simbricks/simbricks/blob/d76fc0eacfedef051ac3fe555950d9c34c1e32a5/lib/simbricks/nicbm/nicbm.cc#L598) part of `nicbm.cc` file. 
Basically this part does what I have mentioned above.
Polls the queues and find the smallest timestamp(`next_ts` in the code) among the messages received.
and if the smallest timestamp is larger than the simulator's current local time, it updates the local time.

Hope it helped.

**Comment by FreakyPenguin** [comment]
I will close this issue for now. Feel free to re-open if you have follow-up questions.
